tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/websocket/RetryClientTest.java

```diff
@@ -0,0 +1,159 @@
+import org.ballerinalang.test.util.websocket.server.WebSocketRemoteServer;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This Class tests the retry support of the WebSocket client and server.
+ */
+@Test(groups = {"websocket-test"})
+public class RetryClientTest extends WebSocketTestCommons {
+
+    private WebSocketRemoteServer remoteServer;
+    private static final String url =  "ws://localhost:21030";
+    private static final int port = 15300;
+    private static final String sentText = "hi all";
+
+    @Test(description = "Tests the retry function using the WebSocket client (starting the server " +
+            ", sending, and receiving text frames.  After that, restart that server and do the same.)")
+    public void testRetry() throws URISyntaxException, InterruptedException, BallerinaTestException {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        remoteServer = new WebSocketRemoteServer(port);
+        remoteServer.run();
+        WebSocketTestClient client = new WebSocketTestClient(url);
+        client.handshake();
+        client.setCountDownLatch(countDownLatch);
+        client.sendText(sentText);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        Assert.assertEquals(client.getTextReceived(), sentText);
+        remoteServer.stop();
+        CountDownLatch latchForRestart = new CountDownLatch(1);
+        remoteServer.run();
+        latchForRestart.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        CountDownLatch latchForSendData = new CountDownLatch(1);
+        client.setCountDownLatch(latchForSendData);
+        client.sendText(sentText);
+        latchForSendData.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        Assert.assertEquals(client.getTextReceived(), sentText);
+        remoteServer.stop();
+        client.shutDown();
+    }
+
+    @Test(description = "Tests the retry function with the maximum count using the WebSocket client (starting " +
+            "the given server, sending, and receiving text frames. After that, restart that server, sending and" +
+            ", sending and receiving text frames then restart that server and repeat)")
+    public void testMultipleRetryAttempts() throws URISyntaxException, InterruptedException, BallerinaTestException {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        String text = "hi madam";
+        ByteBuffer bufferData = ByteBuffer.wrap(new byte[]{1, 2, 3, 4, 6});
+        remoteServer = new WebSocketRemoteServer(port);
+        remoteServer.run();
+        WebSocketTestClient client = new WebSocketTestClient(url);
+        client.handshake();
+        client.setCountDownLatch(countDownLatch);
+        client.sendText(text);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
```

Comment:
I think we are doing the sending text part in all functions, that is we are duplicating code. Can we move it to a separate function then it will add more clarity. Even if not duplicating we can move to a separate function then we would be able to understand the flow. Suppose if the code looks like the following:

```
sendTextData(data);
restartServer();
sendTextData(data);
restartServer();
sendBinaryData();
```
Don't you think all these long functions would be easy to understand? Shall we fix that?